### PR TITLE
[IE CLDNN] Added ScatterElementsUpdate op support

### PR DESCRIPTION
**Pull request by isanghao**

### CODE REVIEW

**Comment by isanghao** [comment]
@geunhwan  @yeonbok  @sungeunk FYI

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/scatter_update/scatter_elements_update_kernel_ref.cpp [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+    case DataLayout::bfzyx:
+        dispatchData.gws = {scope.X().v * scope.Y().v, scope.Z().v, scope.Feature().v * scope.Batch().v};
+        break;
+
+    case DataLayout::bfwzyx:
+        dispatchData.gws = {scope.X().v * scope.Y().v, scope.Z().v * scope.W().v, scope.Feature().v * scope.Batch().v};
+        break;
+    default:
+        assert(0);
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/scatter_update/scatter_elements_update_kernel_ref.cpp:111]: I think you should throw an exception here to reject this kernel and get some meaningful error

File: inference-engine/src/cldnn_engine/cldnn_primitives_list.hpp [hunk 2]
```diff
@@ -166,7 +166,7 @@ REGISTER_FACTORY(v3, ExtractImagePatches);
 // REGISTER_FACTORY(v3, NonZero);
 // REGISTER_FACTORY(v3, ROIAlign);
 // REGISTER_FACTORY(v3, ReadValue);
-// REGISTER_FACTORY(v3, ScatterElementsUpdate);
+REGISTER_FACTORY(v3, ScatterElementsUpdate);
```
[the remark this answers is not in the document]
Comment (isanghao) [reply inference-engine/src/cldnn_engine/cldnn_primitives_list.hpp]: applied.

File: inference-engine/src/cldnn_engine/ops/scatter_elements_update.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
+// Copyright (C) 2020 Intel Corporation
```
[the remark this answers is not in the document]
Comment (isanghao) [reply inference-engine/src/cldnn_engine/ops/scatter_elements_update.cpp]: applied, thanks.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/scatter_update/scatter_elements_update_kernel_ref.cpp (hunk 1 above)
Comment (isanghao) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/scatter_update/scatter_elements_update_kernel_ref.cpp]: applied, thanks.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/scatter_elements_update_ref.cl [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        output[output_idx] = TO_OUTPUT_TYPE(FUSED_OPS_RESULT_SECOND_KERNEL);
+    #else
+        output[output_idx] = ACTIVATION(val, ACTIVATION_PARAMS);
+    #endif
+#endif
+}
+
+#undef GET_UPDATES_INDEX
+#undef GET_OUTPUT_INDEX
+#undef IDX_ORDER
```
[the remark this answers is not in the document]
Comment (isanghao) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/scatter_elements_update_ref.cl:154]: applied, thanks.

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 5]
```diff
@@ -5743,6 +5744,204 @@ INSTANTIATE_TEST_CASE_P(fusings_gpu, scatter_update_scale_activation,
...
+#define CASE_scatter_elements_update_5D_FP16_2 {1, 13, 1, 2, 1}, {1, 2, 1, 2, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_f, data_types::f16, format::bfzyx, data_types::f16, format::bfzyx
+#define CASE_scatter_elements_update_5D_FP16_3 {2, 3, 1, 13, 13}, {2, 3, 1, 2, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_y, data_types::f16, format::bfzyx, data_types::f16, format::bfzyx
+#define CASE_scatter_elements_update_5D_FP16_4 {3, 2, 2, 2, 25}, {3, 2, 2, 2, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_z, data_types::f16, format::bfzyx, data_types::f16, format::bfzyx
+#define CASE_scatter_elements_update_5D_FP16_5 {1, 1, 4, 5, 5}, {1, 1, 3, 1, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_x, data_types::f16, format::bfzyx, data_types::f16, format::bfzyx
+
+#include  "api/engine.hpp"
+class ScatterElementsUpdatePrimitiveFusingTest : public ::BaseFusingTest<scatter_elements_update_test_params> {
+public:
+    ScatterElementsUpdatePrimitiveFusingTest() {
+        this->engine = cldnn::engine(cldnn::engine_configuration {
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp:5793]: Please remove debug code

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 6]
```diff
@@ -5743,6 +5744,204 @@ INSTANTIATE_TEST_CASE_P(fusings_gpu, scatter_update_scale_activation,
...
+
+#define CASE_scatter_elements_update_5D_FP16_1 {13, 2, 1, 2, 1}, {2, 2, 1, 2, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_b, data_types::f16, format::bfzyx, data_types::f16, format::bfzyx
+#define CASE_scatter_elements_update_5D_FP16_2 {1, 13, 1, 2, 1}, {1, 2, 1, 2, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_f, data_types::f16, format::bfzyx, data_types::f16, format::bfzyx
+#define CASE_scatter_elements_update_5D_FP16_3 {2, 3, 1, 13, 13}, {2, 3, 1, 2, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_y, data_types::f16, format::bfzyx, data_types::f16, format::bfzyx
+#define CASE_scatter_elements_update_5D_FP16_4 {3, 2, 2, 2, 25}, {3, 2, 2, 2, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_z, data_types::f16, format::bfzyx, data_types::f16, format::bfzyx
+#define CASE_scatter_elements_update_5D_FP16_5 {1, 1, 4, 5, 5}, {1, 1, 3, 1, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_x, data_types::f16, format::bfzyx, data_types::f16, format::bfzyx
+
+#include  "api/engine.hpp"
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp:5789]: Is it needed here?

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 7]
```diff
@@ -5743,6 +5744,204 @@ INSTANTIATE_TEST_CASE_P(fusings_gpu, scatter_update_scale_activation,
...
+#define CASE_scatter_elements_update_FP16_4 {6, 2, 1, 1}, {1, 2, 1, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_b, data_types::f16, format::bfyx, data_types::f16, format::bfyx
+#define CASE_scatter_elements_update_FP16_5 {3, 1, 6, 8}, {3, 1, 1, 2}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_y, data_types::f16, format::bfyx, data_types::f16, format::bfyx
+
+#define CASE_scatter_elements_update_5D_FP32_1 {24, 3, 1, 4, 1}, {4, 3, 1, 2, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_b, data_types::f32, format::bfzyx, data_types::f32, format::bfzyx
+#define CASE_scatter_elements_update_5D_FP32_2 {2, 17, 2, 2, 2}, {1, 2, 2, 2, 2}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_f, data_types::f32, format::bfzyx, data_types::f32, format::bfzyx
+#define CASE_scatter_elements_update_5D_FP32_3 {5, 3, 2, 20, 22}, {5, 1, 1, 2, 2}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_y, data_types::f32, format::bfzyx, data_types::f32, format::bfzyx
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp:5779]: nit: other macros use upper case for operation name. I think it should be aligned

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp [hunk 8]
```diff
@@ -5743,6 +5744,204 @@ INSTANTIATE_TEST_CASE_P(fusings_gpu, scatter_update_scale_activation,
...
+// This is not a limitation of operation itself, but a limitation of test implementation.
+#define CASE_scatter_elements_update_FP32_1 {8, 4, 1, 1}, {2, 4, 1, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_b, data_types::f32, format::bfyx, data_types::f32, format::bfyx
+#define CASE_scatter_elements_update_FP32_2 {8, 1, 1, 1}, {4, 1, 1, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_b, data_types::f32, format::bfyx, data_types::f32, format::bfyx
+#define CASE_scatter_elements_update_FP32_3 {4, 3, 1, 1}, {2, 2, 1, 1}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_b, data_types::f32, format::bfyx, data_types::f32, format::bfyx
+#define CASE_scatter_elements_update_FP32_4 {2, 8, 1, 2}, {2, 2, 1, 2}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_f, data_types::f32, format::bfyx, data_types::f32, format::bfyx
+#define CASE_scatter_elements_update_FP32_5 {2, 3, 10, 10}, {2, 2, 1, 2}, cldnn::scatter_elements_update::scatter_elements_update_axis::along_y, data_types::f32, format::bfyx, data_types::f32, format::bfyx
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp:5769]: Do these test cases cover different fusion scenarios? At first glance, we can reduce test cases count here without coverage loss

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp (hunk 5 above)
Comment (isanghao) [reply inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp]: Oops, my mistake. Removed.

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp (hunk 6 above)
Comment (isanghao) [reply inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp]: It was for debugging. Removed.

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp (hunk 8 above)
Comment (isanghao) [reply inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp]: Agree. Reduce the test cases.

File: inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp (hunk 7 above)
Comment (isanghao) [reply inference-engine/thirdparty/clDNN/tests/test_cases/fusings_gpu_test.cpp]: Applied, thanks!

**Comment by isanghao** [comment]
@vladimir-paramuzov  Your comments are applied. In addition, I rebased this patch and squashed into single commit.

**Review by vladimir-paramuzov — APPROVED** [review]

**Pull request merged**